### ISSUE TITLE
 How to localize form builder

### ISSUE THREAD

**Issue by Vedhachala** [formio/formio.js#1639]
Hi, 
We are using the angular wrapper for formio (form builder in specific)
Formiojs version : 3.21.5

I want to use form builder in other languages. It would help if you can tell the way to add i18n translations for form builder.

Thanks,
Vedhachala

**Comment by randallknutson** [comment]
http://formio.github.io/formio.js/app/examples/language.html

**Comment by Vedhachala** [comment]
@randallknutson

 I am able to do this for a form where i give the json and the translations for the strings in it. I followed the steps given [here](https://github.com/formio/angular-formio/wiki/Translations) and it works.

But i want to do the same to the builder , which is rendered by this code `<form-builder [form]="{
  components: []
}"></form-builder>`

To be specific, examples of strings i want to give my translations are: 

'Drag and Drop a form component'
.
'Basic Components' (tabs on the left)
'Text Field'
.
'Display' (In edit dialog)
'Validation'
'Logic'
.
'Custom Default Value'
'Calculated Value'

....etc

basically for all the components under basic, advanced, layout and data
also for the input fields (like 'label','Description','Calculated Value','Custom Default Value') in the edit dialog box under display, data, validation, api, conditional, logic tabs for every component.

How to achieve this ?

**Comment by randallknutson** [comment]
These are supposed to be able to be translated the same as other values. However, I took a quick look and it appears we are missing the translation wrapper around some of these strings. For example, the "Drag and Drop a form component" is located here: https://github.com/formio/formio.js/blob/master/src/templates/bootstrap/builderPlaceholder/form.ejs

and should be wrapped in `{{t('Drag and Drop a form component')}}` so that it can be translatable. To fix this, we can either add these wrappers where they are needed or you can override the templates. You can see https://github.com/formio/formio.js/wiki/Form-Templating for how to do this. 

The Items like Display, Validation and others like that should already be translatable the same way as Forms.

**Comment by Lastik** [comment]
How about DateTimeComponent (form renderer and viewer)? I was unable to figure out how to translate it. As I understand It wraps flatpickr, and flatpickr itself already has translation of some (not all) of the string constants. I was unable to make it localised through angular API. Some hint would be great!

**Comment by Vedhachala** [comment]
@randallknutson 

I was able to do localization,
But i am seeing few issues , the contents inside all the code inputs (like in custom default value and caluclate value) are not getting changed.
![uploadFormioIssues1](https://user-images.githubusercontent.com/15833345/64593645-0a3bef80-d3cc-11e9-8a34-3d0c0170fd0e.png)

How to translate this section .?

**Comment by aiwebb** [comment]
@Vedhachala, that text does not currently support localization, but you can certainly override the function that renders it: https://github.com/formio/formio.js/blob/76ef892261c7c6ca12fc450a34cb94f7ce464029/src/components/_classes/component/editForm/utils.js#L38

**Comment by Vedhachala** [comment]
@aiwebb ,
 I went through the code, to examine how to translate a string using i18n. 

I can see that it was just wrapped around a i18next function , for example `this.t(this.component.description)`

However, I am not able to use that function in utils , at least not by simply importing BaseComponent in the utils.

Can you give any suggestions of ways to use it in editForm utils ?

formiojs version : 3.21.5

**Comment by wag110894** [comment]
@Vedhachala, 

Form.io offers Support Packages that are for training, support or development services. The hours can be used in one hour increments at your convenience over a 90 day period.  We will assign one of our senior developers to work directly with you in online sessions. 

To let us know if you are interested, contact support@form.io and we will send you our Support pricing document.

Thanks.

**Comment by Lastik** [comment]
If anyone is interested in DateTimeComponent localization, this could be done the following way:
The standard way of flatpickr localization doesn't work because of the following unresolved issue in flatpickr
https://github.com/flatpickr/flatpickr/issues/1429

Workaround:
1) Include flatpickr to package.json dependencies:
"flatpickr": "^4.6.2"
2) Add the following code to e.g. app.module.ts
(paths may vary based on the project structure and required language):
require("./../../node_modules/flatpickr/dist/l10n/ru.js");
(<any>window).flatpickr.localize((<any>window).flatpickr.l10ns.ru);

**Comment by swapnil147** [comment]
@Vedhachala I am also looking for the same localization on component properties.
Looking forward to detailed information on how did you make it work, what are the changes you made to achieve it?

thank you...

**Comment by danielzzz** [comment]
One question, do you know if there is a file with all the string to be translated? I have been searching for it in the code but I haven't seen it

**Comment by travist** [comment]
@danielzzz most of the translations for this library use these strings.  

https://github.com/formio/formio.js/blob/master/src/translations/en.js

There may be some missing there, and if so, then we welcome a pull request to resolve some of the missing strings.

**Comment by danielzzz** [comment]
thanks, @travist ! from what I've seen lot's of strings are being taken dynamically from components. I was looking for a way to dynamically extract them, but didn't found one yet
